src/components.ts

```diff
@@ -7,336 +7,379 @@
  *
  */
 
-var util = require('racer/lib/util');
-var derbyTemplates = require('./templates');
-var templates = derbyTemplates.templates;
-var expressions = derbyTemplates.expressions;
-var Controller = require('./Controller').Controller;
-var slice = [].slice;
-
-exports.Component = Component;
-exports.ComponentAttribute = ComponentAttribute;
-exports.ComponentAttributeBinding = ComponentAttributeBinding;
-exports.ComponentFactory = ComponentFactory;
-exports.SingletonComponentFactory = SingletonComponentFactory;
-exports.createFactory = createFactory;
-exports.extendComponent = extendComponent;
-
-function Component(context, data) {
-  var parent = context.controller;
-  var id = context.id();
-  var scope = ['$components', id];
-  var model = parent.model.root.eventContext(id);
-  model._at = scope.join('.');
-  data.id = id;
-  model._set(scope, data);
-  // Store a reference to the component's scope such that the expression
-  // getters are relative to the component
-  model.data = data;
-
-  Controller.call(this, parent.app, parent.page, model);
-  this.parent = parent;
-  this.context = context.componentChild(this);
-  this.id = id;
-  this._scope = scope;
-
-  // Add reference to this component on the page so that all components
-  // associated with a page can be destroyed when the page transitions
-  this.page._components[id] = this;
-  this.isDestroyed = false;
+import { type Model, type ModelData } from 'racer';
+import util = require('racer/lib/util');
+
+import { Controller } from './Controller';
+import { PageBase } from './Page';
+import derbyTemplates = require('./templates');
+import { Context } from './templates/contexts';
+import { Expression } from './templates/expressions';
+import { Attribute, Binding } from './templates/templates';
+const { expressions, templates } = derbyTemplates;
+const slice = [].slice;
+
+export interface DataConstructor extends Record<string, unknown> {
+  new(): Record<string, unknown>
 }
 
-util.mergeInto(Component.prototype, Controller.prototype);
+export interface ComponentConstructor extends Component {
+  new(context?: Context, data?: ModelData): Component;
+  DataConstructor?: DataConstructor;
+}
 
-Component.prototype.destroy = function() {
-  this.emit('destroy');
-  this.model.removeContextListeners();
-  this.model.destroy();
-  delete this.page._components[this.id];
-  if (this.page._eventModel.object) {
-    var components = this.page._eventModel.object.$components;
-    if (components) delete components.object[this.id];
+export abstract class Component extends Controller {
+  context: Context;
+  id: string;
+  init?: (model: Model) => void;
+  isDestroyed: boolean;
+  page: PageBase;
+  parent: Controller;
+  singleton?: true;
+  _scope: string[];
+  // new style view prop
+  view?: {
```

Comment:
Not necessarily blocking for this PR as long as we're not publishing the types, but something to think about...

This is used as a static property, never as an instance property, e.g.
```
class MyComponent {
  static view = { is: 'my-component' };
}
```

So we ideally should be defining it as a static property, if possible. That may allow us to do `typeof Component` instead of using `ComponentConstructor` as well?